File: anna/matcher.js
// anna/matcher.js
// ─────────────────────────────────────────────────────────────────────────────
// The matching algorithm — deterministic, explainable, testable (no API key).
//
// Layer 1: HARD GATES (pass/fail). Any FAIL => INELIGIBLE for that carrier.
//          A gate whose driver data is missing is UNKNOWN, not FAIL — the
//          carrier stays a candidate "pending data" instead of being rejected
//          for information we simply haven't collected yet.
// Layer 2: SOFT SCORE (0..100). Among carriers a driver passes, rank by fit.
//
// Claude is NOT used here. Anna uses the LLM only to NORMALIZE messy input into
// the structured DriverProfile (see normalize.js); the scoring itself is plain
// arithmetic so every decision is auditable ("failed gate mvr.maxDUI: 1 > 0").
// ─────────────────────────────────────────────────────────────────────────────

import { OPS, NUMERIC_OPS, getPath, compileSpec } from './spec.js';

export const STATUS = { ELIGIBLE: 'ELIGIBLE', NEEDS_DATA: 'NEEDS_DATA', INELIGIBLE: 'INELIGIBLE' };
const GATE = { PASS: 'PASS', FAIL: 'FAIL', UNKNOWN: 'UNKNOWN' };

const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : null);
const clamp01 = (x) => Math.max(0, Math.min(1, x));

/**
 * Evaluate one driver against one compiled spec's hard gates.
 * @returns {{ status, gateResults, failed, unknown }}
 */
export function evaluateGates(spec, driver) {
  const gateResults = spec.hardGates.map((g) => {
    const dv = getPath(driver, g.field);
    // Missing data for a numeric comparison => UNKNOWN (needs the record/doc).
    if ((dv == null || dv === '') && NUMERIC_OPS.has(g.op)) {
      return { id: g.id, category: g.category, status: GATE.UNKNOWN, reason: `${g.label}: no data on file yet.` };
    }
    const fn = OPS[g.op];
    const pass = fn ? fn(dv, g.value) : true;
    return {
      id: g.id,
      category: g.category,
      status: pass ? GATE.PASS : GATE.FAIL,
      reason: pass ? `${g.label}: OK.` : explainFail(g, dv),
    };
  });

  const failed = gateResults.filter((r) => r.status === GATE.FAIL);
  const unknown = gateResults.filter((r) => r.status === GATE.UNKNOWN);
  let status = STATUS.ELIGIBLE;
  if (failed.length) status = STATUS.INELIGIBLE;
  else if (unknown.length) status = STATUS.NEEDS_DATA;

  return { status, gateResults, failed, unknown };
}

function explainFail(g, dv) {
  const have = dv == null || dv === '' ? 'none on file' : Array.isArray(dv) ? dv.join(', ') || 'none' : dv;
  switch (g.op) {
    case 'lte': return `${g.label}: has ${have}, exceeds the limit of ${g.value}.`;
    case 'gte': return `${g.label}: has ${have}, below the required ${g.value}.`;
    case 'cdlRankGte': return `${g.label}: holds ${have ? 'Class ' + have : 'no class'}.`;
    case 'includesAll': {
      const missing = (g.value || []).filter((x) => !(Array.isArray(dv) ? dv : []).includes(x));
      return `${g.label}: missing ${missing.join(', ')}.`;
    }
    case 'notOwnerOperator': return `${g.label}: applicant is an owner-operator.`;
    default: return `${g.label}: does not meet requirement (${have}).`;
  }
}

/**
 * Soft fit score (0..100) for a driver the gates allow. Only the factors that
 * have both a carrier requirement and driver data contribute; weights are
 * re-normalized over the contributing factors so a sparse spec still scores 0..100.
 */
export function scoreSoft(spec, driver) {
  const reqs = spec.raw || {};
  const w = spec.softWeights;
  const factors = {}; // factorId -> { weight, value(0..1) }

  // Experience margin: up to +5 years above the minimum is "full marks".
  const minExp = num(reqs.cdl?.minExperienceYears);
  const exp = num(driver.cdl?.experienceYears);
  if (minExp != null && exp != null) factors.experienceMargin = { weight: w.experienceMargin, value: clamp01((exp - minExp) / 5) };

  // Clean-record margin: average headroom under MVR/PSP caps (0 = at the cap, 1 = spotless).
  const headrooms = [
    margin(driver.mvr?.movingViolations, reqs.mvr?.maxMovingViolations),
    margin(driver.mvr?.accidents, reqs.mvr?.maxAccidents),
    margin(driver.mvr?.dui, reqs.mvr?.maxDUI),
    margin(driver.psp?.crashes, reqs.psp?.maxCrashes),
    margin(driver.psp?.oosInspections, reqs.psp?.maxOOSInspections),
  ].filter((x) => x != null);
  if (headrooms.length) factors.cleanRecordMargin = { weight: w.cleanRecordMargin, value: headrooms.reduce((a, b) => a + b, 0) / headrooms.length };

  // Endorsements match: fraction of requested endorsements the driver holds.
  const wanted = reqs.cdl?.endorsements || [];
  if (wanted.length) {
    const have = driver.cdl?.endorsements || [];
    factors.endorsementsMatch = { weight: w.endorsementsMatch, value: clamp01(wanted.filter((e) => have.includes(e)).length / wanted.length) };
  }

  // CDL validity margin: comfortably valid beyond the minimum (cap at +365 days).
  const minValid = num(reqs.cdl?.minValidityDays);
  const expiresIn = num(driver.cdl?.expiresInDays);
  if (minValid != null && expiresIn != null) factors.cdlValidityMargin = { weight: w.cdlValidityMargin, value: clamp01((expiresIn - minValid) / 365) };

  const ids = Object.keys(factors);
  if (!ids.length) return { score: 0, breakdown: {} };
  const wsum = ids.reduce((s, id) => s + factors[id].weight, 0) || 1;
  let score = 0;
  const breakdown = {};
  for (const id of ids) {
    const contribution = (factors[id].weight / wsum) * factors[id].value;
    breakdown[id] = Math.round(factors[id].value * 100);
    score += contribution;
  }
  return { score: Math.round(score * 100), breakdown };
}

// Headroom under a "max" cap, normalized 0..1 (1 = zero incidents, 0 = at the cap).
function margin(actual, cap) {
  const a = num(actual), c = num(cap);
  if (a == null || c == null) return null;
  if (c <= 0) return a <= 0 ? 1 : 0;
  return clamp01((c - a) / c);
}

/**
 * Plain-language "why this carrier fits this driver" summary — the offer rationale
 * shown in the driver profile. Positive framing for ELIGIBLE/NEEDS_DATA; for an
 * INELIGIBLE carrier the gate reasons already explain why not, so returns null.
 */
export function buildFitSummary(spec, driver, status, score) {
  if (status === STATUS.INELIGIBLE) return null;
  const reqs = spec.raw || {};
  const pts = [];

  const exp = num(driver.cdl?.experienceYears), minExp = num(reqs.cdl?.minExperienceYears);
  if (exp != null && minExp != null) pts.push(`${exp} yr${exp === 1 ? '' : 's'} experience vs ${minExp} required`);
  else if (exp != null) pts.push(`${exp} yr${exp === 1 ? '' : 's'} experience`);

  const incidents = [driver.mvr?.movingViolations, driver.mvr?.accidents, driver.mvr?.dui, driver.psp?.crashes, driver.psp?.oosInspections].filter((x) => x != null);
  if (incidents.length && incidents.every((x) => Number(x) === 0)) pts.push('clean driving & safety record');
  else if (incidents.length) pts.push('violations within this carrier\'s limits');

  const wanted = reqs.cdl?.endorsements || [];
  if (wanted.length) {
    const have = driver.cdl?.endorsements || [];
    if (wanted.every((e) => have.includes(e))) pts.push(`holds required endorsement${wanted.length > 1 ? 's' : ''} (${wanted.join(', ')})`);
  }
  if (num(driver.cdl?.expiresInDays) != null && num(reqs.cdl?.minValidityDays) != null) pts.push(`CDL valid ${driver.cdl.expiresInDays} days`);
  if (num(driver.age) != null && num(reqs.eligibility?.minAge) != null) pts.push(`meets the ${reqs.eligibility.minAge}+ age requirement`);

  const lead = status === STATUS.ELIGIBLE ? (score >= 80 ? 'Strong fit' : 'Qualifies') : 'Likely fit, pending records';
  const tail = status === STATUS.NEEDS_DATA ? ' MVR/PSP/Clearinghouse not pulled yet.' : '';
  return pts.length ? `${lead} for ${spec.carrierName} — ${pts.join(', ')}.${tail}` : `${lead} for ${spec.carrierName}.${tail}`;
}

/**
 * Match ONE driver against MANY carriers and rank the results.
 *
 * @param {Object} driver   Normalized DriverProfile (see normalize.js).
 * @param {Object[]} carriers  Carrier records (with .requirements) OR pre-compiled specs.
 * @param {Object} [opts]
 * @param {number} [opts.minScore=0]  Drop ELIGIBLE carriers below this fit score.
 * @returns {{ matches, top, summary }}
 */
export function matchDriver(driver, carriers = [], opts = {}) {
  const specs = carriers.map((c) => (c.hardGates ? c : compileSpec(c)));
  const matches = specs.map((spec) => {
    const { status, gateResults, failed, unknown } = evaluateGates(spec, driver);
    const { score, breakdown } = status === STATUS.INELIGIBLE ? { score: 0, breakdown: {} } : scoreSoft(spec, driver);
    return {
      carrierId: spec.carrierId,
      carrierName: spec.carrierName,
      specVersion: spec.version,
      status,
      fitScore: score,
      scoreBreakdown: breakdown,
      gateResults,
      reasons: [...failed, ...unknown].map((r) => r.reason),
      fitSummary: buildFitSummary(spec, driver, status, score),
      // "near miss" => ineligible on a single gate; useful for re-match suggestions.
      nearMiss: status === STATUS.INELIGIBLE && failed.length === 1 ? failed[0].reason : null,
    };
  });

  // Rank: ELIGIBLE (by fit desc) > NEEDS_DATA (fewest unknowns) > INELIGIBLE (near-misses first).
  const order = { [STATUS.ELIGIBLE]: 0, [STATUS.NEEDS_DATA]: 1, [STATUS.INELIGIBLE]: 2 };
  matches.sort((a, b) => {
    if (order[a.status] !== order[b.status]) return order[a.status] - order[b.status];
    if (a.status === STATUS.ELIGIBLE) return b.fitScore - a.fitScore;
    if (a.status === STATUS.NEEDS_DATA) return countUnknown(a) - countUnknown(b);
    return (b.nearMiss ? 1 : 0) - (a.nearMiss ? 1 : 0);
  });

  const min = num(opts.minScore) || 0;
  const eligible = matches.filter((m) => m.status === STATUS.ELIGIBLE && m.fitScore >= min);
  const top = eligible[0] || matches.find((m) => m.status === STATUS.NEEDS_DATA) || null;

  return {
    matches,
    top,
    summary: {
      eligible: eligible.length,
      needsData: matches.filter((m) => m.status === STATUS.NEEDS_DATA).length,
      ineligible: matches.filter((m) => m.status === STATUS.INELIGIBLE).length,
      nearMisses: matches.filter((m) => m.nearMiss).map((m) => ({ carrierId: m.carrierId, carrierName: m.carrierName, reason: m.nearMiss })),
    },
  };
}

const countUnknown = (m) => m.gateResults.filter((r) => r.status === 'UNKNOWN').length;

/**
 * Suggest carriers to re-match a rejected driver to (Stage 5).
 * Excludes carriers already rejected; returns eligible carriers ranked by fit,
 * then near-miss carriers (one gate away) as "stretch" options.
 */
export function suggestRematch(driver, carriers, { excludeCarrierIds = [] } = {}) {
  const { matches } = matchDriver(driver, carriers);
  const ex = new Set(excludeCarrierIds);
  const eligible = matches.filter((m) => m.status === STATUS.ELIGIBLE && !ex.has(m.carrierId));
  const stretch = matches.filter((m) => m.nearMiss && !ex.has(m.carrierId));
  return { eligible, stretch };
}

File: service/anna/matcher.py
"""Matching engine — Python port of anna/matcher.js. Deterministic + explainable."""
from __future__ import annotations
from .spec import OPS, NUMERIC_OPS, get_path, compile_spec, num

STATUS = {"ELIGIBLE": "ELIGIBLE", "NEEDS_DATA": "NEEDS_DATA", "INELIGIBLE": "INELIGIBLE"}
GATE = {"PASS": "PASS", "FAIL": "FAIL", "UNKNOWN": "UNKNOWN"}


def _clamp01(x):
    return max(0.0, min(1.0, x))


def _explain_fail(g, dv):
    have = "none on file" if dv in (None, "") else (", ".join(dv) or "none" if isinstance(dv, list) else dv)
    op = g["op"]
    if op == "lte":
        return f"{g['label']}: has {have}, exceeds the limit of {g['value']}."
    if op == "gte":
        return f"{g['label']}: has {have}, below the required {g['value']}."
    if op == "cdlRankGte":
        return f"{g['label']}: holds {('Class ' + str(have)) if have else 'no class'}."
    if op == "includesAll":
        missing = [x for x in (g["value"] or []) if x not in (dv or [])]
        return f"{g['label']}: missing {', '.join(missing)}."
    if op == "notOwnerOperator":
        return f"{g['label']}: applicant is an owner-operator."
    return f"{g['label']}: does not meet requirement ({have})."


def evaluate_gates(spec, driver):
    results = []
    for g in spec["hardGates"]:
        dv = get_path(driver, g["field"])
        if (dv is None or dv == "") and g["op"] in NUMERIC_OPS:
            results.append({"id": g["id"], "category": g["category"], "status": GATE["UNKNOWN"],
                            "reason": f"{g['label']}: no data on file yet."})
            continue
        fn = OPS.get(g["op"])
        ok = fn(dv, g["value"]) if fn else True
        results.append({"id": g["id"], "category": g["category"],
                        "status": GATE["PASS"] if ok else GATE["FAIL"],
                        "reason": f"{g['label']}: OK." if ok else _explain_fail(g, dv)})
    failed = [r for r in results if r["status"] == GATE["FAIL"]]
    unknown = [r for r in results if r["status"] == GATE["UNKNOWN"]]
    status = STATUS["ELIGIBLE"]
    if failed:
        status = STATUS["INELIGIBLE"]
    elif unknown:
        status = STATUS["NEEDS_DATA"]
    return {"status": status, "gateResults": results, "failed": failed, "unknown": unknown}


def _margin(actual, cap):
    a, c = num(actual), num(cap)
    if a is None or c is None:
        return None
    if c <= 0:
        return 1.0 if a <= 0 else 0.0
    return _clamp01((c - a) / c)


def score_soft(spec, driver):
    reqs = spec.get("raw") or {}
    w = spec["softWeights"]
    factors = {}
    cdl, mvr, psp = driver.get("cdl") or {}, driver.get("mvr") or {}, driver.get("psp") or {}

    min_exp, exp = num((reqs.get("cdl") or {}).get("minExperienceYears")), num(cdl.get("experienceYears"))
    if min_exp is not None and exp is not None:
        factors["experienceMargin"] = (w["experienceMargin"], _clamp01((exp - min_exp) / 5))

    rmvr, rpsp = reqs.get("mvr") or {}, reqs.get("psp") or {}
    headrooms = [m for m in [
        _margin(mvr.get("movingViolations"), rmvr.get("maxMovingViolations")),
        _margin(mvr.get("accidents"), rmvr.get("maxAccidents")),
        _margin(mvr.get("dui"), rmvr.get("maxDUI")),
        _margin(psp.get("crashes"), rpsp.get("maxCrashes")),
        _margin(psp.get("oosInspections"), rpsp.get("maxOOSInspections")),
    ] if m is not None]
    if headrooms:
        factors["cleanRecordMargin"] = (w["cleanRecordMargin"], sum(headrooms) / len(headrooms))

    wanted = (reqs.get("cdl") or {}).get("endorsements") or []
    if wanted:
        have = cdl.get("endorsements") or []
        factors["endorsementsMatch"] = (w["endorsementsMatch"], _clamp01(len([e for e in wanted if e in have]) / len(wanted)))

    min_valid, expires = num((reqs.get("cdl") or {}).get("minValidityDays")), num(cdl.get("expiresInDays"))
    if min_valid is not None and expires is not None:
        factors["cdlValidityMargin"] = (w["cdlValidityMargin"], _clamp01((expires - min_valid) / 365))

    if not factors:
        return {"score": 0, "breakdown": {}}
    wsum = sum(wt for wt, _ in factors.values()) or 1
    score = 0.0
    breakdown = {}
    for fid, (wt, val) in factors.items():
        score += (wt / wsum) * val
        breakdown[fid] = round(val * 100)
    return {"score": round(score * 100), "breakdown": breakdown}


def build_fit_summary(spec, driver, status, score):
    if status == STATUS["INELIGIBLE"]:
        return None
    reqs = spec.get("raw") or {}
    cdl, mvr, psp = driver.get("cdl") or {}, driver.get("mvr") or {}, driver.get("psp") or {}
    pts = []
    exp, min_exp = num(cdl.get("experienceYears")), num((reqs.get("cdl") or {}).get("minExperienceYears"))
    if exp is not None and min_exp is not None:
        pts.append(f"{int(exp)} yr{'' if exp == 1 else 's'} experience vs {int(min_exp)} required")
    elif exp is not None:
        pts.append(f"{int(exp)} yr{'' if exp == 1 else 's'} experience")
    incidents = [x for x in [mvr.get("movingViolations"), mvr.get("accidents"), mvr.get("dui"),
                             psp.get("crashes"), psp.get("oosInspections")] if x is not None]
    if incidents and all(num(x) == 0 for x in incidents):
        pts.append("clean driving & safety record")
    elif incidents:
        pts.append("violations within this carrier's limits")
    wanted = (reqs.get("cdl") or {}).get("endorsements") or []
    if wanted and all(e in (cdl.get("endorsements") or []) for e in wanted):
        pts.append(f"holds required endorsement{'s' if len(wanted) > 1 else ''} ({', '.join(wanted)})")
    if num(cdl.get("expiresInDays")) is not None and num((reqs.get("cdl") or {}).get("minValidityDays")) is not None:
        pts.append(f"CDL valid {int(num(cdl.get('expiresInDays')))} days")
    if num(driver.get("age")) is not None and num((reqs.get("eligibility") or {}).get("minAge")) is not None:
        pts.append(f"meets the {int(num((reqs.get('eligibility') or {}).get('minAge')))}+ age requirement")
    lead = ("Strong fit" if score >= 80 else "Qualifies") if status == STATUS["ELIGIBLE"] else "Likely fit, pending records"
    tail = " MVR/PSP/Clearinghouse not pulled yet." if status == STATUS["NEEDS_DATA"] else ""
    name = spec["carrierName"]
    return f"{lead} for {name} — {', '.join(pts)}.{tail}" if pts else f"{lead} for {name}.{tail}"


def match_driver(driver, carriers, opts=None):
    opts = opts or {}
    specs = [c if c.get("hardGates") else compile_spec(c) for c in carriers]
    matches = []
    for spec in specs:
        ev = evaluate_gates(spec, driver)
        sc = {"score": 0, "breakdown": {}} if ev["status"] == STATUS["INELIGIBLE"] else score_soft(spec, driver)
        matches.append({
            "carrierId": spec["carrierId"], "carrierName": spec["carrierName"], "specVersion": spec["version"],
            "status": ev["status"], "fitScore": sc["score"], "scoreBreakdown": sc["breakdown"],
            "gateResults": ev["gateResults"],
            "reasons": [r["reason"] for r in ev["failed"] + ev["unknown"]],
            "fitSummary": build_fit_summary(spec, driver, ev["status"], sc["score"]),
            "nearMiss": ev["failed"][0]["reason"] if ev["status"] == STATUS["INELIGIBLE"] and len(ev["failed"]) == 1 else None,
        })
    order = {STATUS["ELIGIBLE"]: 0, STATUS["NEEDS_DATA"]: 1, STATUS["INELIGIBLE"]: 2}

    def count_unknown(m):
        return len([r for r in m["gateResults"] if r["status"] == "UNKNOWN"])

    def sort_key(m):
        if m["status"] == STATUS["ELIGIBLE"]:
            return (order[m["status"]], -m["fitScore"])
        if m["status"] == STATUS["NEEDS_DATA"]:
            return (order[m["status"]], count_unknown(m))
        return (order[m["status"]], 0 if m["nearMiss"] else 1)

    matches.sort(key=sort_key)
    min_score = num(opts.get("minScore")) or 0
    eligible = [m for m in matches if m["status"] == STATUS["ELIGIBLE"] and m["fitScore"] >= min_score]
    top = eligible[0] if eligible else next((m for m in matches if m["status"] == STATUS["NEEDS_DATA"]), None)
    return {
        "matches": matches, "top": top,
        "summary": {
            "eligible": len(eligible),
            "needsData": len([m for m in matches if m["status"] == STATUS["NEEDS_DATA"]]),
            "ineligible": len([m for m in matches if m["status"] == STATUS["INELIGIBLE"]]),
            "nearMisses": [{"carrierId": m["carrierId"], "carrierName": m["carrierName"], "reason": m["nearMiss"]}
                           for m in matches if m["nearMiss"]],
        },
    }


def suggest_rematch(driver, carriers, exclude_carrier_ids=None):
    ex = set(exclude_carrier_ids or [])
    matches = match_driver(driver, carriers)["matches"]
    return {
        "eligible": [m for m in matches if m["status"] == STATUS["ELIGIBLE"] and m["carrierId"] not in ex],
        "stretch": [m for m in matches if m["nearMiss"] and m["carrierId"] not in ex],
    }
